apps/prairielearn/src/lib/authz-data.ts

```diff
@@ -1,156 +1,211 @@
+/**
+ * If course_id is not provided, but course_instance_id is,
+ * the function will use the course_id from the course instance.
+ */
+async function selectAuthzData({
   user_id,
   course_id,
   course_instance_id,
-  is_administrator,
-  allow_example_course_override,
   ip,
   req_date,
-  req_mode,
-  req_course_role,
-  req_course_instance_role,
 }: {
   user_id: string;
   course_id: string | null;
   course_instance_id: string | null;
-  is_administrator: boolean;
-  allow_example_course_override: boolean;
   ip: string | null;
   req_date: Date;
-  req_mode: string | null;
-  req_course_role: string | null;
-  req_course_instance_role: string | null;
 }) {
   return sqldb.queryOptionalRow(
     sql.select_authz_data,
     {
       user_id,
       course_id,
       course_instance_id,
-      is_administrator,
-      allow_example_course_override,
       ip,
       req_date,
-      req_mode,
-      req_course_role,
-      req_course_instance_role,
     },
     FullAuthzDataSchema,
   );
 }
 
 /**
- * Builds the authorization data for a user on a page.
+ * Builds the authorization data for a user on a page. The optional parameters are used for effective user overrides,
+ * most scenarios should not need to change these parameters.
  *
  * @param params
- * @param params.authn_user - The authenticated user.
- * @param params.course_id - The ID of the course.
+ * @param params.user - The authenticated user.
+ * @param params.course_id - The ID of the course. Inferred from the course instance if null.
  * @param params.course_instance_id - The ID of the course instance.
- * @param params.is_administrator - Whether the user is an administrator.
  * @param params.ip - The IP address of the request.
  * @param params.req_date - The date of the request.
- * @param params.req_mode - The mode of the request.
+ * @param params.overrides - The overrides to apply to the authorization data.
+ * @param params.overrides.is_administrator - Whether the user is an administrator.
+ * @param params.overrides.req_mode - The mode to request.
+ * @param params.overrides.req_course_role - The course role to request.
+ * @param params.overrides.req_course_instance_role - The course instance role to request.
+ * @param params.overrides.allow_example_course_override - Whether to allow overriding the course role for example courses.
  */
-export async function buildAuthzData({
-  authn_user,
+export async function calculateAuthData({
+  user,
   course_id,
   course_instance_id,
-  is_administrator,
   ip,
   req_date,
-  req_mode = null,
+  overrides,
 }: {
-  authn_user: User;
+  user: User;
   course_id: string | null;
   course_instance_id: string | null;
-  is_administrator: boolean;
   ip: string | null;
   req_date: Date;
-  req_mode?: string | null;
-}) {
+  overrides: {
+    is_administrator: boolean;
+    req_mode?: EnumMode;
+    req_course_role?: EnumCourseRole;
+    req_course_instance_role?: EnumCourseInstanceRole;
```

Comment:
Do we need a callout that these are unvalidated, or should we add a `dangerous` prefix to these, or something like that? Or do you think the fact that this is under "overrides" makes it clear enough that this isn't inherently safe?

Put differently: if I call `calculateAuthData` for some user that's a `Viewer` in a course, and I pass in `req_course_role: 'Owner'`, this function will happily tell me that that user is an owner, is that correct?

Reply:
This is the "requested" role for an override. Maybe? It's under overrides, which to me implies that it's dangerous. Maybe drop the `req` prefix?

Reply:
If this were only called from middleware where we're doing all these checks, I wouldn't care too much. But this is being built explicitly as a function that can be called by anyone at any time. I'd rather be extra careful, including making names sufficiently scary.

> Maybe drop the `req` prefix?

I do think that might be a good improvement regardless: it's ambiguous (what does `req` mean), and if you interpret it as "requested", to me that implies that can answer could be "request denied", which is not the case here.